txnsync/msgbuffers_test.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/algorand/go-algorand/test/partitiontest"
+)
+
+// TestMessageBuffersPool tests that a buffer pool can be retrieved and has proper length/capacity properties
+func TestMessageBuffersPool(t *testing.T) {
+
+	counter := 0
+
+	oldFunc := msgBuffersPool.New
+
+	msgBuffersPool = sync.Pool{
+		New: func() interface{} {
+			counter++
+			return oldFunc()
+		},
+	}
+
+	partitiontest.PartitionTest(t)
+
+	for i := 0; i < 10; i++ {
+		bytes := getMessageBuffer()
+		require.Equal(t, 0, len(bytes))
+		require.GreaterOrEqual(t, cap(bytes), messageBufferDefaultInitialSize)
+
+		releaseMessageBuffer(bytes)
+		require.Equal(t, 1, counter)
```

Comment:
this statement is not guaranteed to be correct ( and, in fact, it fails on circle-ci : https://app.circleci.com/pipelines/github/algorand/go-algorand/1365/workflows/e240c3f9-79a7-4f6d-b24d-cf9d008fd1c5/jobs/9631/tests )

The memory might be returned to the pool, but would not be available right away.
another issue with this test is that it modifies the global msgBuffersPool; this modification is made without any mutex, and this could lead to data races.

instead of conducting the test in that way, I would suggest the following:
instead of using getMessageBuffer/releaseMessageBuffer, start by allocating a buffer of size 400 ( just saying ), and fill it up with some "special value". then call releaseMessageBuffer which would place the buffer into the message buffer. 
Subsequent calls should retrieve that buffer, which could be identified by its size and content. ( i.e. you'll need to resize the buffer to it's max capacity ).
last, because you can't get a guarantee that the buffer wouldn't be consumed by other concurrent running tests, you should probably run the test for a while, and verify that you get the said buffer at least once.

Reply:
I've added a new version of the test that does a variation of this.